Merge the Updated Codebase

Linardos opened the pull request.

sarthakpati reviewed (commented):
@psfoley do you see any major changes preventing us from merging this ASAP?

sarthakpati commented on Task_2/README.md:
Instead of completely removing this, I would perhaps just suggest you leave a single comment about Task_2 being available on this link: https://github.com/FeTS-AI/Challenge/tree/2022/Task_2

sarthakpati commented on Task_1/fets_challenge.egg-info/PKG-INFO:
I think this entire folder (`fets_challenge.egg-info`) should be removed.

psfoley commented:
> @psfoley do you see any major changes preventing us from merging this ASAP?

@sarthakpati Nope - I would expect this can be merged without issue.

sarthakpati commented:
> > @psfoley do you see any major changes preventing us from merging this ASAP?
> 
> @sarthakpati Nope - I would expect this can be merged without issue.

Cool, thanks for the confirmation. 

@Linardos: thank you once again for the herculean effort that you and @kta-intel have put together to modernize this codebase. Once you fix the comments that I have, I can start putting together some CI/CD stuff to make our lives much easier.

Linardos commented:
@sarthakpati I addressed your comments. You may merge. Note that a version of generate_predictions for classification (i.e. inference for classification) still needs work to function though. I can do a separate PR for that when I have it

sarthakpati commented:
There are some unresolved conversations, and I am unable to merge until they are resolved. Also, does this PR contain #198 as well?

Linardos commented:
Seems #198 is an outdated part of the PR that I have since then removed because csv files are generated on-the-go. I resolved the conversations. I also made the code default to segmentation which is functional through both training and inference. Classification will still need some work in terms of inferring the metrics, but we can do that in a later PR.

sarthakpati commented on .gitignore:
I think this file needs to be present (otherwise, folks can accidentally add in a *lot* of binary files).

sarthakpati commented on Task_1/FeTS_Challenge.py:
```diff
@@ -518,15 +518,18 @@ def FedAvgM_Selection(local_tensors,
 # this fact, excluding the three hausdorff measurements. As hausdorff distance is 
 # expensive to compute, excluding them will speed up your experiments.
-include_validation_with_hausdorff=True
+include_validation_with_hausdorff=True # IF SET TO FALSE WHOLE CODE RUNS SEGMENTATION FOR SOME REASON
```
This is a bug, right? So I think it makes sense to open an issue to keep track of it.

Linardos replied:
there is already an open issue for it: https://github.com/FeTS-AI/Challenge/issues/200

sarthakpati commented on Task_1/README.md:
````diff
@@ -20,18 +20,18 @@ Please ask any additional questions in our discussion pages on our github site a
-10. ```pip install .``` 
+7. Install Pytorch (2.3.1) for your system:
+   ```pip install torch==2.3.1 torchvision==0.18.1```
+*Note all previous versions of pytorch can be found in these instructions:[https://pytorch.org/get-started/previous-versions/](https://pytorch.org/get-started/previous-versions/)
````
```suggestion
*Note all previous versions of pytorch can be found in [this link](https://pytorch.org/get-started/previous-versions/#v231).
```

sarthakpati commented on Task_1/setup.py:
```diff
@@ -28,11 +28,11 @@
     install_requires=[
-        'openfl @ git+https://github.com/intel/openfl.git@f4b28d710e2be31cdfa7487fdb4e8cb3a1387a5f',
-        'GANDLF @ git+https://github.com/CBICA/GaNDLF.git@e4d0d4bfdf4076130817001a98dfb90189956278',
+        'openfl @ git+https://github.com/securefederatedai/openfl.git@kta-intel/fets-2024-patch-1',
```
This seems to be problematic:

![image](https://github.com/user-attachments/assets/eaceab9d-f15d-4b9f-ae5b-2244ea34c1d8)

Is there any plan to merge this with openfl's `develop` branch so that we can use that instead? This seems prone to breaking, right? Tagging @kta-intel for guidance.

kta-intel replied:
I just opened a PR for it: https://github.com/securefederatedai/openfl/pull/1141
Once it is merged, we can delete that branch and update the `setup.py` to use the `develop` branch directly

Couple notes:
1. I moved the conditional into `loader_gandlf.py` instead of `runner_gandalf.py` since it is logically a better spot to check if we are running in inference mode, rather than setting another. Simple move that tests fine on my end, but be mindful that it doesn't affect anything else on your end
2. The native API (i.e. `import openfl.native as fx`) will likely be deprecated at some point in this or an upcoming release, so it might be a good idea to pin to this specific openfl commit rather than `develop` as a whole

kta-intel replied:
It has been merged

Point 2 is still being discussed. Considering the importance of the FeTS challenge, let me open up the discussion and propose a way to preserve it at least for v1.7

sarthakpati replied:
> propose a way to preserve it at least for v1.7

This would be the ideal solution. And even until **at least** 1.9 (considering the conventions of semantic versioning).

sarthakpati commented on Task_1/setup.py:
```diff
@@ -28,11 +28,11 @@
-        'openfl @ git+https://github.com/intel/openfl.git@f4b28d710e2be31cdfa7487fdb4e8cb3a1387a5f',
-        'GANDLF @ git+https://github.com/CBICA/GaNDLF.git@e4d0d4bfdf4076130817001a98dfb90189956278',
+        'openfl @ git+https://github.com/securefederatedai/openfl.git@kta-intel/fets-2024-patch-1',
+        'GANDLF @ git+https://github.com/CBICA/GaNDLF.git@0.1.0',
```
```suggestion
        'GANDLF @ git+https://github.com/mlcommons/GaNDLF.git@0.1.0',
```

Can we open an issue where we add a CI run on both GaNDLF and OpenFL master/develop branches respectively?

sarthakpati replied:
This is still unresolved.

sarthakpati commented on Task_1/utils/inspect_pickled_model.py:
```diff
@@ -0,0 +1,22 @@
+import pickle
+
+# Path to the pickle file
+pickle_file_path = '/home/locolinux2/.local/workspace/checkpoint/experiment_109/best_model.pkl'
```
This path (and anything similar) is going to be problematic. I will suggest using something that defined based on [`pathlib.Path.cwd()`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.cwd) (or something similar) to ensure non-breakage.

sarthakpati commented on Task_1/FeTS_Challenge.py:
```diff
@@ -518,15 +518,18 @@ def FedAvgM_Selection(local_tensors,
-institution_split_csv_filename = 'small_split.csv'
+#institution_split_csv_filename = 'partitioning_1.csv'
+institution_split_csv_filename = '/home/locolinux2/datasets/MICCAI_FeTS2022_TrainingData/partitioning_2.csv' 
+institution_split_csv_filename = '/home/locolinux2/datasets/MICCAI_FeTS2022_TrainingData/sanity_partitioning.csv' # a small subset for sanity checks and debugging. Comment out to run the actual challenge partition.
 
 # change this to point to the parent directory of the data
-brats_training_data_parent_dir = '/raid/datasets/FeTS22/MICCAI_FeTS2022_TrainingData'
+brats_training_data_parent_dir = '/home/locolinux2/datasets/MICCAI_FeTS2022_TrainingData'
+brats_training_data_parent_dir = '/home/locolinux2/datasets/MICCAI_FeTS2022_Resized'
```
These paths (and anything similar) is going to be problematic. I will suggest using something that defined based on [`pathlib.Path.cwd()`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.cwd) (or something similar) to ensure non-breakage.

sarthakpati commented on Task_2/README.md:
```diff
@@ -1,74 +1 @@
-- User inside singularity containers isn't root: This can lead to `PermissionError` when reading files from the file system like model checkpoints. Make sure that all files that need to be read from inside the container can be read by *all users*, either before copying them in the Dockerfile or adding chmod commands to the Dockerfile.
-
-Any other Errors ? Feel free to contact us: [forum](https://www.synapse.org/#!Synapse:syn28546456/discussion/default)
+Task 2 is still available on this link: https://github.com/FeTS-AI/Challenge/tree/2022/Task_2
```
```suggestion
Information related to Task 2 of the FeTS Challenge is available on this link: https://github.com/FeTS-AI/Challenge/tree/2022/Task_2
```

sarthakpati commented:
Hey @Linardos, apologies for the late review (as you know, I have been swamped). Anyway, great work on all the changes. I am requesting a few minor points of consideration before we can merge this to be used by others. Thanks!

sarthakpati commented:
@Linardos since the PR from @kta-intel has been merged (https://github.com/securefederatedai/openfl/pull/1141), can you please update the setup appropriately?

sarthakpati commented on Task_1/setup.py:
```diff
@@ -28,7 +28,7 @@
     include_package_data=True,
     install_requires=[
-        'openfl @ git+https://github.com/securefederatedai/openfl.git@kta-intel/fets-2024-patch-1',
+        'openfl @ git+https://github.com/securefederatedai/openfl.git@84cff481b1ec8c92fc88c759ab5b8b6f950f9ca4',
```
@kta-intel, is there any plan to incorporate your branch [[ref](https://github.com/securefederatedai/openfl/tree/kta-intel/fets-2024-patch-1)] into OpenFL develop?

It is still showing as unmerged as of today:

![image](https://github.com/user-attachments/assets/95728722-286d-49fe-b1a7-fbeb3728acac)

I think it makes sense for us to merge this PR as-is now (so that the rest of the code is up-to-date), but do a separate PR to point openfl to the latest tag (perhaps along with a simple unit test). Thoughts?

sarthakpati commented on Task_1/utils/inspect_pickled_model.py:
```diff
@@ -1,7 +1,9 @@
+from pathlib import Path
+
+# Construct path relative to current working directory
+pickle_file_path = Path.cwd() / '.local' / 'workspace' / 'checkpoint' / 'experiment_109' / 'best_model.pkl'
```
Are we 100% sure this path will _always_ exist?
